src/modules/musig/main_impl.h

```diff
@@ -150,286 +160,303 @@ int secp256k1_musig_pubkey_tweak_add(const secp256k1_context* ctx, secp256k1_mus
+    secp256k1_ge_set_gej(&summed_nonces[1], &summed_noncesj[1]);
+    if (!secp256k1_musig_compute_noncehash(noncehash, summed_nonces, combined_pk32, msg)) {
         return 0;
     }
+    /* combined_nonce = summed_nonces[0] + b*summed_nonces[1] */
+    secp256k1_scalar_set_b32(&b, noncehash, NULL);
+    secp256k1_ecmult(ecmult_ctx, &combined_noncej, &summed_noncesj[1], &b, NULL);
+    secp256k1_gej_add_ge(&combined_noncej, &combined_noncej, &summed_nonces[0]);
+    secp256k1_ge_set_gej(combined_nonce, &combined_noncej);
+    return 1;
+}
+
+int secp256k1_musig_process_nonces(const secp256k1_context* ctx, secp256k1_musig_session_cache *session_cache, secp256k1_musig_template *sig_template, int *nonce_parity, const unsigned char * const* pubnonces, size_t n_pubnonces, const unsigned char *msg32, const secp256k1_xonly_pubkey *combined_pk, const secp256k1_musig_pre_session *pre_session, const secp256k1_pubkey *adaptor) {
+    secp256k1_gej summed_nonces[2];
+    secp256k1_ge combined_nonce;
+    unsigned char combined_pk32[32];
+    secp256k1_scalar s;
 
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_context_is_built(&ctx->ecmult_ctx));
+    ARG_CHECK(session_cache != NULL);
+    ARG_CHECK(sig_template != NULL);
+    ARG_CHECK(pubnonces != NULL);
+    ARG_CHECK(n_pubnonces > 0);
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(combined_pk != NULL);
+    ARG_CHECK(pre_session != NULL);
+    ARG_CHECK(pre_session->magic == pre_session_magic);
+
+    if (!secp256k1_xonly_pubkey_serialize(ctx, combined_pk32, combined_pk)) {
+        return 0;
+    }
+
+    /* Compute combined nonce and store in sig template */
+    secp256k1_musig_sum_nonces(ctx, summed_nonces, pubnonces, n_pubnonces);
     /* Add public adaptor to nonce */
     if (adaptor != NULL) {
-        secp256k1_pubkey_load(ctx, &noncep, adaptor);
-        secp256k1_gej_add_ge_var(&combined_noncej, &combined_noncej, &noncep, NULL);
+        secp256k1_ge adaptorp;
+        if (!secp256k1_pubkey_load(ctx, &adaptorp, adaptor)) {
+            return 0;
+        }
+        secp256k1_gej_add_ge_var(&summed_nonces[0], &summed_nonces[0], &adaptorp, NULL);
+    }
+    if (!secp256k1_musig_process_nonces_internal(&ctx->ecmult_ctx, &combined_nonce, &session_cache->data[0], summed_nonces, combined_pk32, msg32)) {
+        return 0;
+    }
+    if (!secp256k1_xonly_ge_serialize(&sig_template->data[0], &combined_nonce)) {
+        /* unreachable with overwhelming probability */
+        return 0;
     }
-
     /* Negate nonce if Y coordinate is not square */
-    secp256k1_ge_set_gej(&combined_noncep, &combined_noncej);
-    secp256k1_fe_normalize_var(&combined_noncep.y);
-    session->combined_nonce_parity = secp256k1_extrakeys_ge_even_y(&combined_noncep);
+    secp256k1_fe_normalize_var(&combined_nonce.y);
+    /* Store nonce parity in session cache */
+    session_cache->data[64] = secp256k1_extrakeys_ge_even_y(&combined_nonce);
```

Comment:
Since `combined_nonce` is unused later, I would rather call `session_cache->data[64] =secp256k1_fe_is_odd(combined_nonce.y)` since it does not modify `combined_nonce`

Reply:
Good catch, the negation was unnecessary.